compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -629,8 +629,18 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
 
             // When `async_fn_in_dyn_trait` is enabled, we don't need to check the
             // RPITIT for compatibility, since it's not provided by the user.
-            if tcx.features().async_fn_in_dyn_trait() && tcx.is_impl_trait_in_trait(assoc_type) {
-                continue;
+            if tcx.is_impl_trait_in_trait(assoc_type) {
+                if tcx.features().async_fn_in_dyn_trait() {
+                    continue;
+                } else {
+                    tcx.dcx().span_delayed_bug(
+                        obligation.cause.span,
+                        "RPITIT in trait object shouldn't have been considered",
+                    );
+                    return Err(SelectionError::TraitDynIncompatible(
+                        trait_predicate.trait_ref.def_id,
+                    ));
```

Comment:
is this motivated by a test? Or just being more robust than going through the rest of the logic here?